metadata-ingestion/src/datahub/ingestion/source/sac/sac.py

```diff
@@ -0,0 +1,808 @@
+            token_endpoint=config.token_url,
+            token_endpoint_auth_method="client_secret_post",
+            grant_type="client_credentials",
+        )
+
+        self.session.register_compliance_hook(
+            "protected_request", _add_sap_sac_custom_auth_header
+        )
+        self.session.fetch_token()
+
+        self.client = pyodata.Client(
+            url=f"{self.config.tenant_url}/api/v1",
+            connection=self.session,
+            config=pyodata.v2.model.Config(retain_null=True),
+        )
+
+    def close(self) -> None:
+        self.session.close()
+        super().close()
+
+    @classmethod
+    def create(cls, config_dict: dict, ctx: PipelineContext) -> "SACSource":
+        config = SACSourceConfig.parse_obj(config_dict)
+        return cls(config, ctx)
+
+    @staticmethod
+    def test_connection(config_dict: dict) -> TestConnectionReport:
+        test_report = TestConnectionReport()
+
+        try:
+            config = SACSourceConfig.parse_obj(config_dict)
+
+            session = OAuth2Session(
+                client_id=config.client_id,
+                client_secret=config.client_secret.get_secret_value(),
+                token_endpoint=config.token_url,
+                token_endpoint_auth_method="client_secret_post",
+                grant_type="client_credentials",
+            )
+
+            session.register_compliance_hook(
+                "protected_request", _add_sap_sac_custom_auth_header
+            )
+            session.fetch_token()
+
+            response = session.get(url=f"{config.tenant_url}/api/v1/$metadata")
+            response.raise_for_status()
+
+            response = session.get(url=f"{config.tenant_url}/api/v1/dataimport/models")
+            response.raise_for_status()
+
+            test_report.basic_connectivity = CapabilityReport(capable=True)
+
+        except Exception as e:
+            logger.error(f"Failed to test connection due to {e}", exc_info=e)
+            if test_report.basic_connectivity is None:
+                test_report.basic_connectivity = CapabilityReport(
+                    capable=False, failure_reason=f"{e}"
+                )
+            else:
+                test_report.internal_failure = True
+                test_report.internal_failure_reason = f"{e}"
+
+        return test_report
+
+    def get_workunit_processors(self) -> List[Optional[MetadataWorkUnitProcessor]]:
+        return [
+            *super().get_workunit_processors(),
+            partial(
+                auto_incremental_lineage,
+                self.config.incremental_lineage,
+            ),
+            StaleEntityRemovalHandler.create(
+                self, self.config, self.ctx
+            ).workunit_processor,
+        ]
+
+    def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+        if self.config.ingest_stories or self.config.ingest_applications:
+            resources = self.get_resources()
+
+            for resource in resources:
```

Comment:
it'd be good to pull some of these out into helper methods - the `get_workunits_internal` method is a bit long and hence hard to follow

Reply:
Should be better now?